app/src/main/java/com/example/mycomifclient/connexion/ConnexionActivity.kt

```diff
@@ -1,22 +1,178 @@
+        serverBody.addProperty("password", password)
+        return serverBody
+    }
+
+    private fun authenticate(authBody: JsonObject) {
+
+        retrofitHTTPServices.authenticate(authBody).enqueue(object : Callback<JsonObject> {
+            override fun onResponse(call: Call<JsonObject>, response: Response<JsonObject>) {
+                when (response.raw().code()) {
+
+                    200 -> handleAuthenticationResponse(response.body())
+
+                    401 -> Toast.makeText(
+                        this@ConnexionActivity,
+                        "Wrong credentials, please try again",
+                        Toast.LENGTH_LONG
+                    ).show()
+
+                    else -> println("Error")
+                }
+            }
+
+            override fun onFailure(call: Call<JsonObject>, t: Throwable) {
+                Toast.makeText(this@ConnexionActivity, "Error: $t", Toast.LENGTH_LONG).show()
+            }
+        })
+    }
+
+    private fun getUser() {
+        val user = userDAO.getFirst()
+        retrofitHTTPServices.getUser(218, "Bearer " + user.token)
+            .enqueue(object : Callback<JsonObject> {
+                override fun onResponse(
+                    call: Call<JsonObject>,
+                    response: Response<JsonObject>
+                ) {
+                    when (response.raw().code()) {
+
+                        200 -> handleGetUserResponse(response.body())
+
+                        401 -> reconnect()
+
+                        else -> println("Error")
+                    }
+                }
+
+                override fun onFailure(call: Call<JsonObject>, t: Throwable) {
+                    Toast.makeText(this@ConnexionActivity, "Error: $t", Toast.LENGTH_LONG).show()
+                }
+            })
+    }
+
+    private fun handleAuthenticationResponse(body: JsonObject?) {
+        val token = body?.get("access_token")
+        if (token == null) {
+            Toast.makeText(
+                this,
+                "Error while recovering data from server. Please contact an administrator",
+                Toast.LENGTH_LONG
+            ).show()
+        } else {
+            userDAO.insert(UserEntity(1, "", "", id, password, removeQuotes(token), 0))
+            getUser()
+        }
+    }
+
+    private fun handleGetUserResponse(body: JsonObject?) {
+        if (body != null) {
+            val user = userDAO.getFirst()
+            val userEntity = UserEntity(
+                body.get("id").asInt,
+                removeQuotes(body.get("first_name")),
+                removeQuotes(body.get("last_name")),
+                removeQuotes(body.get("email")),
+                user.password,
+                user.token,
+                body.get("balance").asInt
+            )
+            userDAO.nukeTable()
+            userDAO.insert(userEntity)
+            val intent = Intent(this, MainActivity::class.java)
+            this.startActivity(intent)
```

Comment:
Adding a this.finish() should be useful in order to avoid to go back to this activity when clicking back button in MainActivity.

Reply:
Added 😉